### PR TITLE
MPI parallelization of thermodynamics

### PR DESCRIPTION
**Pull request by draenog**
## Background 
This PR is part of #120, where the basis of strategy for MPI parallelisation is described.

## Change description
As all thermodynamics operations are local to a grid cell all required MPI communication, is handled by NetCDF4 library. Therefore the only required steps are:

- [x] Initialize and finalize MPI stack
- [x] Read decomposition metadata on each rank
- [x] Read and write the necessary part of the grid on each rank (depends on #330)
- [x] Tests for parallel I/O 

**NOTE** PR #432 should be merged into this branch before it is merged into develop 

To run `run_simple_example.sh` you will need to generate the following netcdf file

```
$ ncdump partition.nc 
netcdf partition {
dimensions:
        P = 1 ;
        L = 1 ;
        globalX = 30 ;
        globalY = 30 ;

group: bounding_boxes {
  variables:
        int global_x(P) ;
        int global_y(P) ;
        int local_extent_x(P) ;
        int local_extent_y(P) ;
  data:

   global_x = 0 ;

   global_y = 0 ;

   local_extent_x = 30 ;

   local_extent_y = 30 ;
  } // group bounding_boxes
}

```

### CODE REVIEW

**Comment by dorchard** [comment]
`develop` branch now in a position for this to be turned into a non-draft PR, once the conflicts are resolved, which @draenog will do next week.

**Comment by TomMelt** [comment]
Based on discussion with Tim and Einar, this should be rebased onto `feat_mar3`. I will then look to rebase #265 onto this (`mpi_local`) branch.

**Comment by einola** [comment]
We need to be careful not to use MPP_WORLD as the communicator because the model should receive the communicator from OASIS in a coupled mode.

**Comment by draenog** [comment]
> We need to be careful not to use MPP_WORLD as the communicator because the model should receive the communicator from OASIS in a coupled mode.

Yes, it is a good point. I think it is currently quite well encapsulated. Constructor of  Nextsim:Model takes MPI communicator as an argument which is then saved in ModelMetadata.

**Comment by TomMelt** [comment]
I have rebased this branch to `develop` to that it is ready to be merged in. First however, we should merge #432 into this branch. For some reason github is showing the incorrect diff. I think maybe the GUI is broken. The actual diff is quite small:

```
$ git ds mpi_local
 core/src/RectGridIO.cpp               |  42 +++++++--------
 core/test/CMakeLists.txt              |  30 +++++++++++
 core/test/MPI_example_tests.cpp       |  46 ++++++++++++++++
 core/test/MainMPI.cpp                 |  19 +++++++
 core/test/RectGrid_test.cpp           |  24 +++++++++
 lib/doctest/doctest.h                 |  33 ++++++++----
 lib/doctest/extensions/doctest_mpi.h  | 169 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 lib/doctest/extensions/doctest_util.h |  37 +++++++++++++
 lib/doctest/extensions/mpi_reporter.h | 271 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 lib/doctest/extensions/mpi_sub_comm.h |  84 +++++++++++++++++++++++++++++
 10 files changed, 724 insertions(+), 31 deletions(-)

```

File: core/src/RectGridIO.cpp [hunk 1]
```diff
@@ -22,13 +23,31 @@
...
+    ModelArray::MultiDim dims;
+    dims.resize(nDims);
+    dims[0] = metadata.localExtentX;
+    dims[1] = metadata.localExtentY;
+    for (size_t d = 2; d < nDims; ++d) {
+        dims[d] = dataGroup.getVar(fieldName).getDim(d).getSize();
+    }
+    ModelArray::setDimensions(type, dims);
+}
+#else
```
Comment (timspainNERSC) [inline core/src/RectGridIO.cpp:50]: Is this still correct in a post-dimension ordering swap world?

File: core/src/StructureFactory.cpp [hunk 2]
```diff
@@ -41,25 +42,42 @@ std::string structureNameFromFile(const std::string& filePath)
...
     // TODO There must be a better way
     if (DevGrid::structureName == structureName) {
         Module::setImplementation<IStructure>("DevGrid");
         DevGrid gridIn;
         gridIn.setIO(new DevGridIO(gridIn));
+#ifdef USE_MPI
+        return gridIn.getModelState(filePath, partitionFile, metadata);
+#else
         return gridIn.getModelState(filePath);
+#endif
```
Comment (timspainNERSC) [inline core/src/StructureFactory.cpp:62]: DevGrid has been removed. Is there a need to rebase? Or will the merge/PR sort this all out?

File: core/src/modules/include/DevGrid.hpp
Comment (timspainNERSC) [inline core/src/modules/include/DevGrid.hpp:1]: DevGrid no longer exists in develop.

**Review by timspainNERSC — CHANGES_REQUESTED** [review]
* Need to check that the changes to DevGrid go through cleanly now that it doesn't exist in `develop`.
* Need to check DimensionSetter() is still correct in the MPI version now that the ordering of the array dimensions has been swapped.

Otherwise, this looks like a god set of MPI changes.

File: core/src/StructureFactory.cpp (hunk 2 above)
Comment (MarionBWeinzierl) [reply core/src/StructureFactory.cpp:62]: I think `DevGrid` here can be replaced by `RectGrid` and it should (hopefully) be fine.

File: core/src/RectGridIO.cpp (hunk 1 above)
Comment (timspainNERSC) [reply core/src/RectGridIO.cpp:50]: I see this is dealt with in #432

**Review by timspainNERSC — APPROVED** [review]

**Pull request merged**